csrc/optimization/move_split_cat.cpp

```diff
@@ -0,0 +1,140 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <optimization/move_split_cat.h>
+
+#include <vector>
+
+#include <expr_simplifier.h>
+#include <fusion.h>
+#include <ir/builder.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ir/utils.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+// Finds the canceling split of `cat` and returns the input TensorView of the
+// split. A split (implemented as multiple `slice`s) and a cat cancel when they
+// work on the same dimension. For example, when
+//
+//   s0 = in[:, :5]
+//   s1 = in[:, 5:]
+//   out = cat([s0, s1], dim=-1)
+//
+// findCancelingSplit(out) returns `in`.
+//
+// Currently, the cat has to immediately follow the split. But we will soon
+// handle patterns like `split ->
+// a_chain_of_unary_ops_that_do_not_resize_the_split_dimension -> cat`.
+TensorView* findCancelingSplit(CatOp* cat) {
+  // nvFuser implements cat as PadOp->CatOp. We first locate these `PadOp`s.
+  std::vector<PadOp*> pads;
+  pads.reserve(cat->inputs().size());
+  for (Val* in : cat->inputs()) {
+    pads.push_back(in->definition()->as<PadOp>());
```

Comment:
nipick: should we instead dynamic_cast and return here when the assumption fails?

`in->definition()->as<PadOp>()`

Reply:
I prefer a hard error to quietly being slow when this assumption breaks. This way, the person breaks that assumption will get notified about this downstream effect. Then they can either bypass this optimization understanding the performance impact or improve this pass so it's useful again. 

If it's UX that you are concerned about, CatOp taking padded inputs is currently [enforced](https://github.com/NVIDIA/Fuser/blob/3c6ead7aa10a35d4b58beb2e27226f0bed4c4b22/csrc/ops/alias.cpp#L634-L635) by our C++ API, so the assumption can only be broken from internal.